common/src/main/java/me/lucko/luckperms/common/storage/dao/cassandra/CassandraDao.java

```diff
@@ -0,0 +1,500 @@
+                        }
+                    }
+                }
+            } catch (Exception e) {
+                e.printStackTrace();
+                plugin.getLog().severe("Error occurred whilst initialising the database.");
+                shutdown();
+            }
+        }
+
+        TypeCodec<UDTValue> codec = CodecRegistry.DEFAULT_INSTANCE.codecFor(keyspace.getUserType("permission"));
+        NodeCodec nodeCodec = new NodeCodec(codec);
+        CodecRegistry.DEFAULT_INSTANCE.register(nodeCodec);
+
+        ACTION_INSERT = session.prepare(this.prefix.apply("INSERT INTO {prefix}actions(time, actor_uuid, actor_name, type, acted_uuid, acted_name, action) VALUES(?, ?, ?, ?, ?, ?, ?)"));
+        ACTION_SELECT_ALL = session.prepare(this.prefix.apply("SELECT * FROM {prefix}actions"));
+
+        USER_SELECT_ALL_UUID = session.prepare(this.prefix.apply("SELECT uuid FROM {prefix}users"));
+        USER_SELECT_PERMISSIONS = session.prepare(this.prefix.apply("SELECT uuid, permissions FROM {prefix}users"));
+        USER_SELECT = session.prepare(this.prefix.apply("SELECT * FROM {prefix}users WHERE uuid=?"));
+        USER_DELETE = session.prepare(this.prefix.apply("DELETE FROM {prefix}users WHERE uuid=?"));
+        USER_INSERT = session.prepare(this.prefix.apply("INSERT INTO {prefix}users(uuid, name, permissions, primaryGroup) VALUES(?, ?, ?, ?)"));
+        USER_RENAME = session.prepare(this.prefix.apply("UPDATE {prefix}users SET name=? WHERE uuid=?"));
+        USER_UPDATE_PERMISSIONS = session.prepare(this.prefix.apply("UPDATE {prefix}users SET permissions=? WHERE uuid=?"));
+
+        GROUP_SELECT_ALL = session.prepare(this.prefix.apply("SELECT * FROM {prefix}groups"));
+        GROUP_SELECT = session.prepare(this.prefix.apply("SELECT * FROM {prefix}groups WHERE name=?"));
+        GROUP_INSERT = session.prepare(this.prefix.apply("INSERT INTO {prefix}groups(name, permissions) VALUES(?, ?)"));
+        GROUP_DELETE = session.prepare(this.prefix.apply("DELETE FROM {prefix}groups WHERE name=?"));
+
+        TRACK_SELECT = session.prepare(this.prefix.apply("SELECT * FROM {prefix}tracks WHERE name=?"));
+        TRACK_SELECT_ALL = session.prepare(this.prefix.apply("SELECT * FROM {prefix}tracks"));
+        TRACK_INSERT = session.prepare(this.prefix.apply("INSERT INTO {prefix}tracks(name, groups) VALUES(?, ?)"));
+        TRACK_DELETE = session.prepare(this.prefix.apply("DELETE FROM {prefix}tracks WHERE name=?"));
+
+        UUID_TO_NAME_SELECT = session.prepare(this.prefix.apply("SELECT name FROM {prefix}uuid_to_name WHERE uuid=?"));
+        NAME_TO_UUID_SELECT = session.prepare(this.prefix.apply("SELECT uuid FROM {prefix}name_to_uuid WHERE name=?"));
+        UUID_TO_NAME_INSERT = session.prepare(this.prefix.apply("INSERT INTO {prefix}uuid_to_name(uuid, name) VALUES(?, ?)"));
+        NAME_TO_UUID_INSERT = session.prepare(this.prefix.apply("INSERT INTO {prefix}name_to_uuid(name, uuid) VALUES(?, ?)"));
```

Comment:
I'm not really an expert on Cassandra, but certainly with the JDBC, it's considered bad practice to cache PreparedStatements in this way, as among other things, it produces race conditions when they're used concurrently.

I assume BoundStatements get around this? Idk, just wanted to check. :P